File: diffwake/diffwake_jax/farm.py
import jax.numpy as jnp
from flax import struct                
from typing import Callable, Dict, Optional, Any

POWER_SETPOINT_DEFAULT = 1.0e12

@struct.dataclass
class Farm:
    layout_x: jnp.ndarray                   
    layout_y: jnp.ndarray                    
    n_turbines: int
    turbine: Any                              

    hub_height: float
    rotor_diameter: float
    TSR: float
    ref_tilt: float
    correct_cp_ct_for_tilt: float
    power_function: Callable = struct.field(pytree_node=False)
    thrust_coefficient_function: Callable = struct.field(pytree_node=False)
    axial_induction_function: Callable = struct.field(pytree_node=False)
    tilt_interp: Optional[Callable] = struct.field(pytree_node=False)
    power_thrust_table: Dict = struct.field(pytree_node=False)

    yaw_angles: Optional[jnp.ndarray] = None      
    tilt_angles: Optional[jnp.ndarray] = None    
    power_setpoints: Optional[jnp.ndarray] = None  

    yaw_angles_sorted: Optional[jnp.ndarray] = None
    tilt_angles_sorted: Optional[jnp.ndarray] = None
    power_setpoints_sorted: Optional[jnp.ndarray] = None
    state: str = "UNINITIALIZED"

    @staticmethod
    def create(
        layout_x: jnp.ndarray,
        layout_y: jnp.ndarray,
        turbine_type_dict: Dict,
        turbine_class: Callable,
        yaw_angles: Optional[jnp.ndarray] = None,
    ) -> "Farm":
        turbine = turbine_class(**turbine_type_dict)
        return Farm(
            layout_x=layout_x,
            layout_y=layout_y,
            n_turbines=layout_x.shape[0],
            turbine=turbine,
            hub_height=turbine.hub_height,
            rotor_diameter=turbine.rotor_diameter,
            TSR=turbine.TSR,
            ref_tilt=turbine.power_thrust_table["ref_tilt"],
            correct_cp_ct_for_tilt=turbine.correct_cp_ct_for_tilt,
            power_function=turbine.power_function,
            thrust_coefficient_function=turbine.thrust_coefficient_function,
            axial_induction_function=turbine.axial_induction_function,
            tilt_interp=turbine.tilt_interp,
            power_thrust_table=turbine.power_thrust_table,
            yaw_angles=yaw_angles if yaw_angles is not None else None,
        )

    """
    def set_yaw_angles(self, yaw_angles: jnp.ndarray, sorted_indices: jnp.ndarray) -> "Farm":
        idx = sorted_indices[:, :, 0, 0]
        yaw_s   = jnp.take_along_axis(yaw_angles,  idx, axis=1)
        return self.replace(yaw_angles=yaw_angles,
                            yaw_angles_sorted=yaw_s)
    """
    def set_yaw_angles_to_ref_yaw(self, n_findex: int) -> "Farm":
        yaw_zero = jnp.zeros((n_findex, self.n_turbines))
        return self.replace(yaw_angles=yaw_zero, yaw_angles_sorted=yaw_zero)

    def set_tilt_to_ref_tilt(self, n_findex: int) -> "Farm":
        tilt_ref = jnp.ones((n_findex, self.n_turbines)) * self.ref_tilt
        return self.replace(tilt_angles=tilt_ref, tilt_angles_sorted=tilt_ref)

    def set_power_setpoints_to_ref_power(self, n_findex: int) -> "Farm":
        psp = jnp.ones((n_findex, self.n_turbines)) * POWER_SETPOINT_DEFAULT
        return self.replace(power_setpoints=psp, power_setpoints_sorted=psp)

    def initialize(self, sorted_indices: jnp.ndarray) -> "Farm":
        B,T,_,_ = sorted_indices.shape
        
        idx = sorted_indices[:, :, 0, 0]

        if self.yaw_angles is None:
            yaw = jnp.zeros((B,T))
        else:
            yaw = self.yaw_angles
        if self.tilt_angles is None:
            tilt = jnp.zeros((B,T))
        else:
            tilt = self.tilt_angles
        yaw_s   = jnp.take_along_axis(yaw,  idx, axis=1)
        tilt_s  = jnp.take_along_axis(tilt, idx, axis=1)
        if self.power_setpoints is not None:
            pset_s  = jnp.take_along_axis(self.power_setpoints, idx, axis=1)
        else:
            pset_s = None
        return self.replace(
            yaw_angles = yaw,
            tilt_angles = tilt,
            yaw_angles_sorted=yaw_s,
            tilt_angles_sorted=tilt_s,
            power_setpoints_sorted=pset_s,
            state="INITIALIZED",
        )

    def finalize(self, unsorted_indices: jnp.ndarray) -> "Farm":
        idx = unsorted_indices[:, :, 0, 0]
        yaw   = jnp.take_along_axis(self.yaw_angles_sorted,  idx, axis=1)
        tilt  = jnp.take_along_axis(self.tilt_angles_sorted, idx, axis=1)
        return self.replace(yaw_angles=yaw, tilt_angles=tilt)

    def expand_farm_properties(
        self, n_findex: int, sorted_coord_idx: jnp.ndarray
    ) -> "Farm":
        expand = lambda x: jnp.broadcast_to(x, (n_findex, x.shape[0]))
        tilt_sort = jnp.take_along_axis(
            expand(jnp.array(self.ref_tilt)), 1, sorted_coord_idx
        )
        return self.replace(tilt_angles_sorted=tilt_sort)

    def calculate_tilt_for_eff_velocities(
        self, v_eff: jnp.ndarray
    ) -> jnp.ndarray:
        if self.tilt_interp is not None:
            return self.tilt_interp(v_eff)
        return jnp.zeros_like(v_eff)

    @property
    def coordinates(self) -> jnp.ndarray:
        return jnp.stack(
            [self.layout_x, self.layout_y, jnp.full_like(self.layout_x, self.hub_height)], axis=-1
        )

File: diffwake/diffwake_julia/farm.jl
const POWER_SETPOINT_DEFAULT = 1.0e12

"""
    Farm

Wind-farm layout and turbine property container.  Mirrors the Flax `Farm`
dataclass from `farm.py`.
"""
Base.@kwdef mutable struct Farm
    layout_x::Vector{Float64}
    layout_y::Vector{Float64}
    n_turbines::Int

    # Turbine properties (scalars, extracted from Turbine)
    hub_height::Float64
    rotor_diameter::Float64
    TSR::Float64
    ref_tilt::Float64
    correct_cp_ct_for_tilt::Bool

    # Function handles
    power_function::Function
    thrust_coefficient_function::Function
    axial_induction_function::Function
    tilt_interp::Union{Nothing, Function}

    power_thrust_table::Dict{String, Any}

    # Turbine struct reference
    turbine::Any

    # Sorted / unsorted angles  (B, T)
    yaw_angles::Union{Nothing, Matrix{Float64}}              = nothing
    tilt_angles::Union{Nothing, Matrix{Float64}}             = nothing
    power_setpoints::Union{Nothing, Matrix{Float64}}         = nothing

    yaw_angles_sorted::Union{Nothing, Matrix{Float64}}       = nothing
    tilt_angles_sorted::Union{Nothing, Matrix{Float64}}      = nothing
    power_setpoints_sorted::Union{Nothing, Matrix{Float64}}  = nothing
end

"""
    create_farm(layout_x, layout_y, turbine_type_dict, turbine_class; yaw_angles=nothing)

Factory — mirrors `Farm.create(...)`.
"""
function create_farm(
    layout_x::Vector{Float64},
    layout_y::Vector{Float64},
    turbine_type_dict::Dict,
    turbine_constructor::Function;
    yaw_angles::Union{Nothing, Matrix{Float64}} = nothing,
)
    turbine = turbine_constructor(turbine_type_dict)

    return Farm(
        layout_x          = layout_x,
        layout_y          = layout_y,
        n_turbines        = length(layout_x),
        hub_height        = turbine.hub_height,
        rotor_diameter    = turbine.rotor_diameter,
        TSR               = turbine.TSR,
        ref_tilt          = turbine.power_thrust_table["ref_tilt"],
        correct_cp_ct_for_tilt = turbine.correct_cp_ct_for_tilt,
        power_function              = turbine.power_function,
        thrust_coefficient_function = turbine.thrust_coefficient_function,
        axial_induction_function    = turbine.axial_induction_function,
        tilt_interp                 = turbine.tilt_interp,
        power_thrust_table          = turbine.power_thrust_table,
        turbine                     = turbine,
        yaw_angles                  = yaw_angles,
    )
end

"""
    initialize_farm!(farm, sorted_indices)

Sort yaw/tilt/power-setpoints by the upstream→downstream ordering in
`sorted_indices` (B, T, Ny, Nz).
"""
function initialize_farm!(farm::Farm, sorted_indices::Array{Int,4})
    B, T, _, _ = size(sorted_indices)
    idx = sorted_indices[:, :, 1, 1]   # (B, T)

    yaw  = farm.yaw_angles  === nothing ? zeros(B, T) : farm.yaw_angles
    tilt = farm.tilt_angles === nothing ? zeros(B, T) : farm.tilt_angles

    yaw_s  = _take_along_axis2_mat(yaw,  idx)
    tilt_s = _take_along_axis2_mat(tilt, idx)

    pset_s = if farm.power_setpoints !== nothing
        _take_along_axis2_mat(farm.power_setpoints, idx)
    else
        nothing
    end

    farm.yaw_angles         = yaw
    farm.tilt_angles        = tilt
    farm.yaw_angles_sorted  = yaw_s
    farm.tilt_angles_sorted = tilt_s
    farm.power_setpoints_sorted = pset_s

    return farm
end

# ── helper: gather along axis 2 for matrices ─────────────────────────────────
function _take_along_axis2_mat(src::Matrix{Float64}, idx::Matrix{Int})
    B, T = size(src)
    out = similar(src)
    for b in 1:B, t in 1:T
        out[b, t] = src[b, idx[b, t]]
    end
    return out
end
